File: prj_akhir/lib/models/payment.dart
// lib/models/payment.dart
import 'booking.dart';
import 'promo.dart';
import 'user.dart';

class Payment {
  final int id;
  final int bookingId;
  final int userId;
  final int? promoId;
  final String invoiceNumber;
  final double baseAmount;
  final double discountAmount;
  final double totalAmount;
  final String paymentMethod;
  final String status;
  final String? proofOfPayment;
  final DateTime? paidAt;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;
  final Booking? booking;
  final Promo? promo;
  final User? user;

  Payment({
    required this.id,
    required this.bookingId,
    required this.userId,
    this.promoId,
    required this.invoiceNumber,
    required this.baseAmount,
    required this.discountAmount,
    required this.totalAmount,
    required this.paymentMethod,
    required this.status,
    this.proofOfPayment,
    this.paidAt,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
    this.booking,
    this.promo,
    this.user,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    double parseToDouble(dynamic value) {
      if (value == null) return 0.0;
      if (value is double) return value;
      if (value is int) return value.toDouble();
      if (value is String) {
        final cleaned = value.replaceAll(RegExp(r'[^0-9.]'), '');
        if (cleaned.isEmpty) return 0.0;
        return double.tryParse(cleaned) ?? 0.0;
      }
      return 0.0;
    }

    return Payment(
      id: json['id'] ?? 0,
      bookingId: json['booking_id'] ?? 0,
      userId: json['user_id'] ?? 0,
      promoId: json['promo_id'],
      invoiceNumber: json['invoice_number'] ?? '',
      baseAmount: parseToDouble(json['base_amount']),
      discountAmount: parseToDouble(json['discount_amount']),
      totalAmount: parseToDouble(json['total_amount']),
      paymentMethod: json['payment_method'] ?? '',
      status: json['status'] ?? 'pending',
      proofOfPayment: json['proof_of_payment'],
      paidAt: json['paid_at'] != null
          ? DateTime.tryParse(json['paid_at'])
          : null,
      notes: json['notes'],
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at']) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at']) ?? DateTime.now()
          : DateTime.now(),
      booking: json['booking'] != null
          ? Booking.fromJson(json['booking'])
          : null,
      promo: json['promo'] != null ? Promo.fromJson(json['promo']) : null,
      user: json['user'] != null ? User.fromJson(json['user']) : null,
    );
  }

  String get statusLabel {
    switch (status) {
      case 'pending':
        return 'Menunggu Verifikasi';
      case 'paid':
        return 'Lunas';
      case 'failed':
        return 'Gagal';
      case 'refunded':
        return 'Dibatalkan';
      default:
        return status;
    }
  }

  String get statusColor {
    switch (status) {
      case 'pending':
        return '#FF9800';
      case 'paid':
        return '#34A853';
      case 'failed':
        return '#EA4335';
      case 'refunded':
        return '#80868B';
      default:
        return '#80868B';
    }
  }

  String get paymentMethodLabel {
    switch (paymentMethod) {
      case 'bank_transfer_bca':
        return 'Bank Transfer BCA';
      case 'bank_transfer_mandiri':
        return 'Bank Transfer Mandiri';
      case 'bank_transfer_bni':
        return 'Bank Transfer BNI';
      case 'ovo':
        return 'OVO';
      case 'gopay':
        return 'GoPay';
      default:
        return paymentMethod;
    }
  }

  bool get hasProof => proofOfPayment != null && proofOfPayment!.isNotEmpty;
  String get proofUrl => proofOfPayment != null
      // ? 'http://192.168.126.112:8000/storage/payments/$proofOfPayment'
      ? 'http://127.0.0.1:8000/storage/payments/$proofOfPayment'
      : '';
}

File: be_prjakhir/app/Models/Payment.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Payment extends Model
{
    protected $fillable = [
        'booking_id',
        'user_id',
        'promo_id',
        'invoice_number',
        'base_amount',
        'discount_amount',
        'total_amount',
        'payment_method',
        'status',
        'proof_of_payment',
        'paid_at',
        'notes'
    ];

    protected $casts = [
        'base_amount' => 'decimal:2',
        'discount_amount' => 'decimal:2',
        'total_amount' => 'decimal:2',
        'paid_at' => 'datetime'
    ];

    // Relasi ke Booking
    public function booking()
    {
        return $this->belongsTo(Booking::class);
    }

    // Relasi ke User
    public function user()
    {
        return $this->belongsTo(User::class);
    }

    // Relasi ke Promo
    public function promo()
    {
        return $this->belongsTo(Promo::class);
    }

    // Accessor untuk payment method label
    public function getPaymentMethodLabelAttribute()
    {
        $methods = [
            'bank_transfer_bca' => 'Bank Transfer BCA',
            'bank_transfer_mandiri' => 'Bank Transfer Mandiri',
            'bank_transfer_bni' => 'Bank Transfer BNI',
            'ovo' => 'OVO',
            'gopay' => 'GoPay'
        ];
        return $methods[$this->payment_method] ?? $this->payment_method;
    }

    // Accessor untuk status label
    public function getStatusLabelAttribute()
    {
        $statuses = [
            'pending' => 'Menunggu Verifikasi',
            'paid' => 'Lunas',
            'failed' => 'Gagal',
            'refunded' => 'Dibatalkan'
        ];
        return $statuses[$this->status] ?? $this->status;
    }

    // Accessor untuk proof URL
    public function getProofUrlAttribute()
    {
        if ($this->proof_of_payment) {
            return asset('storage/payments/' . $this->proof_of_payment);
        }
        return null;
    }

    // Generate invoice number
    public static function generateInvoiceNumber()
    {
        $prefix = 'INV-' . date('Ymd');
        $lastPayment = self::where('invoice_number', 'LIKE', $prefix . '%')
            ->orderBy('id', 'desc')
            ->first();
            
        if (!$lastPayment) {
            return $prefix . '-0001';
        }
        
        $lastNumber = intval(substr($lastPayment->invoice_number, -4));
        $newNumber = str_pad($lastNumber + 1, 4, '0', STR_PAD_LEFT);
        
        return $prefix . '-' . $newNumber;
    }
}